### DISCUSSION TITLE
Information for compilation needed

### DISCUSSION THREAD

**Discussion by peterjit** [jamulussoftware/jamulus#1861] [Documentation]
I wanted to update to version 3.8.0 on my raspberry pi. Unfortunately the [old installation page](https://github.com/jamulussoftware/jamulus/wiki/Installation-for-Linux), where the process of compilation was described has moved and [the new one lacks this information](https://jamulus.io/wiki/Installation-for-Linux) I know this is quite a noob question but I'd be grateful if someone could help me out here...

**Comment by jujudusud** [comment]
You can now read the compiling informations here :
https://github.com/jamulussoftware/jamulus/blob/master/COMPILING.md

**Reply by peterjit** [reply]
Thank you for your reply! I found this page but doesn't it just contain the information how to compile the dependencies and not the software itself? I remember only a few bits like directing to the download folder and then "qmake Jamulus.pro".

**Reply by peterjit** [reply]
I looked into the version history of the installation page and found the information I needed:

## Compile this bad boy

Now `cd` into the jamulus sources directory you downloaded:

```shell
cd jamulus-latest
```
Now compile the sources with the following commands (the last `make` may take several minutes to run):


~~~
qmake Jamulus.pro
make clean
make
sudo make install
~~~


I was able to update to 3.8.0 with these instructions. Maybe there are others who have as little experience as myself and the information should be added to the new page.

**Reply by dcorson-ticino-com** [reply]
Yes this should be somewhere where it can be found.

On the other hand you are safer to never depend on the web, make a local copy and you will always have it when needed.

**Reply by jujudusud** [reply]
I think you should have a deeper look at the compiling.md file. There are all the informations to build the software in it.
Please see standard desktop paragraph.
It is different than the previous version.

**Reply by peterjit** [reply]
Maybe it's my lack of experience, but I really don't see these commands in the compiling.md file. What I found under "Standard desktop build" produced an error message on my raspberry. I was able to install with the old instructions. But maybe this is a problem with my system. I followed @dcorson-ticino-com 's advice and made a local copy of the old page. Thanks for both of your feedback!

**Reply by ghost** [reply]
The Compile this bad boy page should be on the Jamulus webpage somewhere because the instructions are Excellent. I have a friend who compiled and installed Jamulus for Ubuntu by following these instructions after I told him that you type the commands in the command line terminal. That was the first time he compiled anything; not even Hello World. I wasn't even present when he did this.

Furthermore, the instructions on installing Jamulus dependencies etc. on this page are required also.

> I looked into the version history of the installation page and found the information I needed:
> ## Compile this bad boy
> 
> Now `cd` into the jamulus sources directory you downloaded:
> 
> ```shell
> cd jamulus-latest
> ```
> 
> Now compile the sources with the following commands (the last `make` may take several minutes to run):
> 
> ```
> qmake Jamulus.pro
> make clean
> make
> sudo make install
> ```
> 
> I was able to update to 3.8.0 with these instructions. Maybe there are others who have as little experience as myself and the information should be added to the new page.

**Reply by ann0see** [reply]
@gilgongo removed it because the official way is now to use the .deb files. 

Probably we still need to add the old installation guidelines for compilation somewhere else?

**Reply by gilgongo** [reply]
I'm a little confused by this. I didn't remove any of the information as far as I'm aware. Can someone explain what's missing exactly?

**Reply by rdica** [reply]
I'd say yes. I prefer to compile, and there's no RPMs for non-deb based linux distros. I use both Ubuntu and Fedora with Jamulus.

**Reply by peterjit** [reply]
> I'm a little confused by this. I didn't remove any of the information as far as I'm aware. Can someone explain what's missing exactly?

Sorry for the confusion... It's this bit that I needed to compile it for the Raspberry's architecture: 
## Compile this bad boy

Now `cd` into the jamulus sources directory you downloaded:

```shell
cd jamulus-latest
```
Now compile the sources with the following commands (the last `make` may take several minutes to run):
~~~
qmake Jamulus.pro
make clean
make
sudo make install
~~~

**Reply by jujudusud** [reply]
> What I found under "Standard desktop build" produced an error message on my raspberry.

I still haven't received my crystal ball... Can you tell us what the error is? Following which 
Command Line ?
I already have an idea of what's going on.

@gilgongo , I think it is 'make distclean' that causes an error because it wanted to delete /usr/local/ as I read from another Pi user.

**Reply by peterjit** [reply]
> > What I found under "Standard desktop build" produced an error message on my raspberry.
> 
> I still haven't received my crystal ball... Can you tell us what the error is? Following which
> Command Line ?
> I already have an idea of what's going on.
> 
> @gilgongo , I think it is 'make distclean' that causes an error because it wanted to delete /usr/local/ as I read from another Pi user.

Sorry, I can't retrieve this and as it is running now I don't want to risk corrupting the installation. But I'm going to install the new version on the other raspberries of my ensemble end of June. I'll try the new method again and come back to this thread with the error message.

**Reply by gilgongo** [reply]
@peterjit so the error you got was when you ran `make distclean` in the current instructions, is that right? 

Also, are you compiling because the pre-compiled .deb installer doesn't work for you?

@DavidSavinkoff 

> Furthermore, the instructions on installing Jamulus dependencies etc. on this page are required also.

`sudo apt install ./jamulus-[release].deb` does that automatically. With autobuilt binaries, manual compilation for Linux is now not necessary, any more than it is for other platforms, so the documentations was changed accordingly.

**Reply by ghost** [reply]
> With autobuilt binaries, manual compilation for Linux is now not necessary, any more than it is for other platforms, so the documentations was changed accordingly.
> 
Throwing away excellent documentation is a regression. The documentation should be put into an appendix.

**Reply by gilgongo** [reply]
@peterjit  Can you tell us why you are compiling by hand and not using the .deb binaries? It might help other Pi users.

**Reply by jujudusud** [reply]
@DavidSavinkoff Compiling informations are Always in a readme file or compiling in the sources.

**Reply by peterjit** [reply]
@gilgongo 
> @peterjit so the error you got was when you ran `make distclean` in the current instructions, is that right?

Exactly. I tried again, the error reads:
make: *** No rule to make target 'distclean'.  Stop.
I stopped at this point because I remembered that the old documentation that had worked for me was completely different and started to look for it.
> @peterjit Can you tell us why you are compiling by hand and not using the .deb binaries? It might help other Pi users.

Yes. I didn't find an arm64 .deb file for version 3.8.0 but the instructions from the old installation documentation page work perfectly. When I first installed Jamulus 3.7.3 a month ago it was literally my first day with Linux and still I was able to follow the instructions.

I'm well aware that most of these problems probably have to do with my lack of understanding...

**Reply by jujudusud** [reply]
Fyi

`qmake` is used to configure the project for the architecture where you build.
Naming Jamulus.pro after qmake command is not an obligation if there is only one .pro file in the directory.
`make` will build all the file necessary for the app.
`sudo make install` will copy all the files on your system.

**Reply by gilgongo** [reply]
@peterjit 
>I didn't find an arm64 .deb file for version 3.8.0 

Ah OK. So it sounds like we need to amend the compile steps that are currently there to read:

~~~
make distclean # make clean on Raspberry Pi
qmake # qmake-qt5 on Fedora 33
make
sudo make install
~~~

Can anyone confirm? I have no test equipment for this.

**Reply by dcorson-ticino-com** [reply]
I use the following for my ARM headless server:

qmake "CONFIG+=nosound headless"
make
sudo make install

**Reply by jujudusud** [reply]
> I use the following for my ARM headless server:
> 
> qmake "CONFIG+=nosound headless"
> make
> sudo make install

You also have to clean the source if the build as always been done before.
nosound and headless options are relevant only if you want to build an headless server only.
Could you please have a building test with the following:

- qmake
- make clean
- make

And tell us if there is an error somewhere ?

**Reply by dcorson-ticino-com** [reply]
I could do a client build, but can't test the result as my hardware has no head.
It is just a headless server.
I don't do the clean as I remove the sources after making the executable.

**Reply by npostavs** [reply]
`make distclean` works if you have run `qmake` before so that `Makefile` exists (e.g., you built the previous version, and are now doing `git pull` to update). If you download the sources to a new directory every time, then cleaning isn't needed.

>     make: *** No rule to make target 'distclean'. Stop.

This happens if you run `make distclean` without running `qmake` first. Or if you run it twice (because it deletes the `Makefile`). Maybe it's better to suggest `make clean` because it doesn't have that fails-after-the-second-time caveat.

There's nothing raspberry pi specific about it.
